drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -90,236 +113,295 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
 	return 0;
 }
 
-static int bq274xx_read_data_block(const struct device *dev, uint8_t offset,
-				   uint8_t *data, uint8_t bytes)
+static int bq274xx_get_device_type(const struct device *dev, uint16_t *val)
 {
-	const struct bq274xx_config *config = dev->config;
-	uint8_t i2c_data;
 	int ret;
 
-	i2c_data = BQ274XX_EXT_BLKDAT_START + offset;
-
-	ret = i2c_burst_read_dt(&config->i2c, i2c_data, data, bytes);
+	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
 	if (ret < 0) {
-		LOG_ERR("Failed to read block");
+		LOG_ERR("Unable to write control register");
 		return -EIO;
 	}
 
-	k_sleep(BQ274XX_SUBCLASS_DELAY);
+	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL, val);
+	if (ret < 0) {
+		LOG_ERR("Unable to read register");
+		return -EIO;
+	}
 
 	return 0;
 }
 
-static int bq274xx_get_device_type(const struct device *dev, uint16_t *val)
+static int bq274xx_read_block(const struct device *dev,
+			      uint8_t subclass,
+			      uint8_t *block, uint8_t num_bytes)
 {
+	const struct bq274xx_config *const config = dev->config;
 	int ret;
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_DEVICE_TYPE);
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_CLASS, subclass);
 	if (ret < 0) {
-		LOG_ERR("Unable to write control register");
+		LOG_ERR("Failed to update state subclass");
 		return -EIO;
 	}
 
-	ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_CONTROL_LOW, val);
+	/* DataBlock(), 0 for the first 32 bytes. */
+	ret = i2c_reg_write_byte_dt(&config->i2c, BQ274XX_EXT_DATA_BLOCK, 0x00);
 	if (ret < 0) {
-		LOG_ERR("Unable to read register");
+		LOG_ERR("Failed to update block offset");
+		return -EIO;
+	}
+
+	k_sleep(BQ274XX_SUBCLASS_DELAY);
+
+	ret = i2c_burst_read_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
+	if (ret < 0) {
+		LOG_ERR("Unable to read block data");
 		return -EIO;
 	}
 
 	return 0;
 }
 
-static int bq274xx_gauge_configure(const struct device *dev)
+static int bq274xx_write_block(const struct device *dev,
+			       uint8_t *block, uint8_t num_bytes)
 {
 	const struct bq274xx_config *const config = dev->config;
-	struct bq274xx_data *data = dev->data;
-	const struct bq274xx_regs *regs = data->regs;
+	uint8_t checksum = 0;
 	int ret;
-	uint8_t tmp_checksum, checksum_old, checksum_new;
-	uint16_t flags, designenergy_mwh, taperrate, reg_val;
-	uint8_t block[BQ27XXX_DM_SZ];
-	uint8_t try;
 
-	designenergy_mwh = (uint32_t)config->design_capacity * 37 / 10; /* x3.7 */
-	taperrate = config->design_capacity * 10 / config->taper_current;
-
-	/* Unseal the battery control register */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
```

Comment:
Ok, so the issue I was having (nRF52840, I2C compatible = "nordic,nrf-twim") seems to be:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131832/e6251988-1c82-40aa-995b-65b22dfc14e6)

So for best compatibility I'd suggest we change from the `i2c_burst_write_dt` API to `i2c_write_dt`.
```suggestion
	uint8_t buf[1 + BQ27XXX_DM_SZ] = {BQ274XX_EXT_BLKDAT_START};
	__ASSERT_NO_MSG(num_bytes <= BQ27XXX_DM_SZ);
	memcpy(&buf[1], block, num_bytes);
	ret = i2c_write_dt(&config->i2c, buf, 1 + num_bytes);
```

Reply:
no way! the irony, I normally test on nrf52, this time I'm on an stm32 :-( oky switched to your API, many thanks for testing and spotting this. A bit shocking though, few drivers using `i2c_burst_write_dt`.